### PR TITLE
merge metadata object library from ambient

### PR DESCRIPTION
**Pull request by kyessenov**
Signed-off-by: Kuat Yessenov <kuat@google.com>

Remove ssl from metadata object: TLS forwarding is handled by tls_passthrough filter combo.
Merge metadata object into main.

### CODE REVIEW

**Comment by kyessenov** [comment]
/assign @stevenctl 
for confirmation about control plane usage
/assign @douglas-reid 
as code owner

**Comment by kyessenov** [comment]
/assign @briansonnenberg
for C++ readability

File: src/envoy/common/metadata_object_test.cc
Comment (kyessenov) [inline src/envoy/common/metadata_object_test.cc:92]: @douglas-reid this breaks hashing, doesn't it? there's is no instance for deployment.

**Comment by stevenctl** [comment]
Seems fine from CP standpoint.

File: src/envoy/common/metadata_object_test.cc
Comment (kyessenov) [reply src/envoy/common/metadata_object_test.cc:92]: @howardjohn why isn't instance set for deployments?

Comment (howardjohn) [reply src/envoy/common/metadata_object_test.cc:92]: Where is it trying to read from? WorkloadMetadataResource over ECDS?

If so, as far as I can tell we should always have the instance name?

if its from baggage, then I don't think that is included as part of baggage, the format we have is just `k8s.deployment.name=<deployment_name>`

**Comment by kyessenov** [comment]
@zirain Want to take a look?

We're slowly merging the new baggage encoding for peer metadata into main code base.

**Comment by kyessenov** [comment]
@douglas-reid can you approve if you think this is improving your code? There's no impact on istio existing filters.

**Comment by zirain** [comment]
> @zirain Want to take a look?
> 
> We're slowly merging the new baggage encoding for peer metadata into main code base.

Is there any design doc for this?

**Comment by kyessenov** [comment]
@zirain I checked but couldn't find any in the community groups. The gist is very simple, for HTTP CONNECT, we are using a different HTTP header to pass the peer metadata.

**Before:**: Header `x-envoy-peer-metadata`, which is a base64 encoded protobuf struct, closely aligned to schema https://github.com/istio/proxy/blob/master/extensions/common/node_info.fbs

**After**: Header `baggage`, which is a standard HTTP encoding `key1=value1,key2=value2`. The schema is defined in `WorkloadMetadataObject` in this PR.

The primary advantage is standardization on the RFC https://www.w3.org/TR/baggage/.

**Comment by zirain** [comment]
glad to see removal of flatbuff, which's hard to use in my personal experience
I used to think istio use flatbuff because of performace, correct me if I'm wrong
so, after moving to `baggage`, will performace become better or worse?

**Comment by kyessenov** [comment]
Micro-benchmark result (rough, do not assume precise numbers):
*Before*: header length 356, 4456ns
*After*: header length 163, 572ns

So the new encoding is 2x more compact and 10x faster to decode.

File: extensions/common/metadata_object.cc
Comment (zirain) [inline extensions/common/metadata_object.cc:71]: add comments for last two `empty`?

File: extensions/common/metadata_object.cc
Comment (zirain) [inline extensions/common/metadata_object.cc:43]: change these `if-else` to `switch-case`, WDYT?

Comment (kyessenov) [reply extensions/common/metadata_object.cc:43]: You can't switch on strings :) But yes, we can tokenize here with enums, will do that.

File: extensions/common/metadata_object.cc
Comment (zirain) [inline extensions/common/metadata_object.cc:210]: auto owners = absl.StrSplit(owner, "/")];
switch (owners[len-1]){
case 'deployments':
case 'cornjobs':
}

Comment (kyessenov) [reply extensions/common/metadata_object.cc:210]: This would be doing more work since it's a long string. Probably easier to find last index and just tokenize/switch on the last part.

Comment (zirain) [reply extensions/common/metadata_object.cc:210]: avoid multi `absl:EndsWith`

Comment (kyessenov) [reply extensions/common/metadata_object.cc:210]: Yeah, will fix this with switch. Thanks.

File: extensions/common/metadata_object.cc
Comment (kyessenov) [reply extensions/common/metadata_object.cc:106]: Removed those two fields since they are unused.

File: extensions/common/metadata_object.cc
Comment (kyessenov) [reply extensions/common/metadata_object.cc]: Done.

**Comment by kyessenov** [comment]
@zirain I'll need your approval from P&T group since there is no one else around :)

**Comment by zirain** [comment]
sorry, I'm OOO for national holidays. will take a look when BTO.

**Review by zirain — APPROVED** [review]
LGTM, thanks

**Pull request merged**
